blocks/aem-asset-selector/aem-asset-selector.js

```diff
@@ -1,48 +1,113 @@
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
 import { readBlockConfig } from '../../scripts/lib-franklin.js';
 import {
-  init, renderAssetSelectorWithImsFlow, logoutImsFlow,
+  init,
+  renderAssetSelectorWithImsFlow,
+  logoutImsFlow,
+  copyAssetWithoutRapi,
+  copyAssetWithRapi,
 } from './aem-asset-selector-util.js';
 
-export default function decorate(block) {
+export default async function decorate(block) {
   let rendered = false;
+  let selected = false;
   const cfg = readBlockConfig(block);
   block.textContent = '';
   block.innerHTML = `
-    <div class="asset-overlay">
+    <div class="asset-overlay loading">
+      <img id="loading" src="${cfg.loading}" />
       <div id="login">
-        <p>Welcome to the Asset Selector! After signing in you'll have the option to select which assets to use.</p>
+        <p>Welcome to the Asset Selector! Please sign in to view your assets.</p>
         <button id="as-login">Sign In</button>
       </div>
     </div>
     <div class="action-container">
+        <div><img src="${cfg.logo}" /></div>
+        <button id="as-copy" class="disabled">Copy</button>
         <button id="as-cancel">Sign Out</button>
     </div>
-    <dialog id="asset-selector-dialog">
-        <div id="asset-selector" style="height: calc(100vh - 80px); width: calc(100vw - 60px); margin: -20px;">
-        </div>
-    </dialog>
-    `;
+    <div id="asset-selector">
+    </div>
+  `;
+
   block.querySelector('#as-login').addEventListener('click', (e) => {
     e.preventDefault();
     renderAssetSelectorWithImsFlow(cfg);
   });
 
+  const copy = block.querySelector('#as-copy');
+  copy.addEventListener('click', async (e) => {
+    e.preventDefault();
+    if (selected) {
+      copy.classList.add('disabled');
+      copy.innerText = 'Copying...';
+      const copyMethod = cfg['use-rapi'] ? copyAssetWithRapi : copyAssetWithoutRapi;
+      const success = await copyMethod(selected);
+      if (success) {
+        copy.innerText = 'Copied!';
+      } else {
+        copy.innerText = 'Error!';
+      }
+      copy.classList.remove('disabled');
+    }
+  });
+
   block.querySelector('#as-cancel').addEventListener('click', (e) => {
     e.preventDefault();
     logoutImsFlow();
   });
 
+  // give a little time for onAccessTokenReceived() to potentially come in
+  setTimeout(() => {
+    const overlay = block.querySelector('.asset-overlay');
+    if (overlay.style.display !== 'none') {
+      // at this point the overlay is still visible, meaning that we haven't
+      // gotten an event indicating the user is logged in. Display the
+      // sign in interface
+      overlay.classList.remove('loading');
+      block.querySelector('#loading').style.display = 'none';
+      block.querySelector('#login').style.display = 'flex';
+    }
+  }, 2000);
```

Comment:
Make the timeout value a constant, so we can reuse the same wait times for the UI/UX.